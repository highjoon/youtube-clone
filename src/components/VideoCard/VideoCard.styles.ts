import styled from '@emotion/styled';

export const Container = styled.li`
  display: flex;
  flex-direction: column;
  width: 276px;
  gap: 13px;
  cursor: pointer;
`;

export const VideoInfo = styled.div`
  position: relative;
`;

export const VideoThumbnail = styled.div<{ url: string }>`
  height: 153px;
  background-image: url(${props => props.url});
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
`;

export const Duration = styled.div`
  position: absolute;
  right: 5px;
  bottom: 5.5px;
  width: 42px;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 1px 4px;
  background-color: var(--color-primary-black);
  color: var(--color-primary-white);
`;

export const DetailWrapper = styled.div`
  display: flex;
  gap: 14px;
`;

export const ChannelThumbnail = styled.img`
  display: inline;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  border-radius: 50%;
`;

export const MediaInfoWrapper = styled.div``;

export const VideoTitle = styled.span`
  font-size: 14px;
  color: var(--color-primary-black);
  margin-bottom: 5px;
  white-space: pre-wrap;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
`;

export const ChannelTitle = styled.span`
  display: block;
  font-size: 14px;
  color: var(--color-black-dark2);
`;

export const ViewCount = styled.span`
  font-size: 14px;
  line-height: 16px;
  color: var(--color-black-dark2);
`;

export const PublishedAt = styled.span`
  font-size: 14px;
  line-height: 16px;
  color: var(--color-black-dark2);
`;
