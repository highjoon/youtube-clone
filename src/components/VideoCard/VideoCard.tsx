import React from 'react';

import { formatAgo, formatNumber } from '@/utils/format';

import * as Styled from './VideoCard.styles';

interface VideoThumbnailProps extends Props {
  url: string;
}

interface ChannelThumbnailProps extends Props {
  url: string;
}

export const Container = ({ className, children }: PropsWithReactNodeChildren) => {
  const handleClickCard = () => {};

  return (
    <Styled.Container className={className} onClick={handleClickCard}>
      {children}
    </Styled.Container>
  );
};

export const VideoInfo = ({ className, children }: PropsWithReactNodeChildren) => {
  return <Styled.VideoInfo className={className}>{children}</Styled.VideoInfo>;
};

export const VideoThumbnail = ({ className, url }: VideoThumbnailProps) => {
  return <Styled.VideoThumbnail className={className} url={url}></Styled.VideoThumbnail>;
};

export const Duration = ({ className, children }: PropsWithStringChildren) => {
  const content = children.substring(2, children.length - 1).replace('M', ':');

  return <Styled.Duration className={className}>{content}</Styled.Duration>;
};

export const Detail = ({ className, children }: PropsWithReactNodeChildren) => {
  return <Styled.DetailWrapper className={className}>{children}</Styled.DetailWrapper>;
};

export const VideoTitle = ({ className, children }: PropsWithStringChildren) => {
  return <Styled.VideoTitle className={className}>{children}</Styled.VideoTitle>;
};

export const MediaInfo = ({ className, children }: PropsWithReactNodeChildren) => {
  return <Styled.MediaInfoWrapper className={className}>{children}</Styled.MediaInfoWrapper>;
};

export const ChannelThumbnail = ({ className, url }: ChannelThumbnailProps) => {
  return <Styled.ChannelThumbnail className={className} src={url} />;
};

export const ChannelTitle = ({ className, children }: PropsWithStringChildren) => {
  return <Styled.ChannelTitle className={className}>{children}</Styled.ChannelTitle>;
};

export const ViewCount = ({ className, children }: PropsWithStringChildren) => {
  const content = formatNumber(Number(children));

  return <Styled.ViewCount className={className}>{content} views â€¢</Styled.ViewCount>;
};

export const PublishedAt = ({ className, children }: PropsWithStringChildren) => {
  const formattedPublishedAt = new Date(children);
  const content = formatAgo(formattedPublishedAt);

  return <Styled.PublishedAt className={className}> {content}</Styled.PublishedAt>;
};
