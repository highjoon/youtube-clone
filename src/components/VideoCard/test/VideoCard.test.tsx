import { render } from '@testing-library/react';

import VideoCard from '..';

const VideoCardComponent = (
  <VideoCard>
    <VideoCard.VideoInfo>
      <VideoCard.VideoThumbnail
        url={
          'https://blog.snappa.com/wp-content/uploads/2020/01/BestYoutubeThumbnailSize_Blogpost.jpg'
        }
      />
      <VideoCard.Duration>PT4M13S</VideoCard.Duration>
    </VideoCard.VideoInfo>
    <VideoCard.Detail>
      <VideoCard.ChannelThumbnail url={'https://avatars.githubusercontent.com/u/63948484?v=4'} />
      <VideoCard.MediaInfo>
        <VideoCard.VideoTitle>
          {` The Dark Knight Rises (2012) - Batman's Sacrifice Scene (9/10) Batman's Sacrifice Scene
          (9/10)`}
        </VideoCard.VideoTitle>
        <VideoCard.ChannelTitle>Movieclips Trailers</VideoCard.ChannelTitle>
        <VideoCard.ViewCount>60000000</VideoCard.ViewCount>
        <VideoCard.PublishedAt>2023-01-12T06:06:22Z</VideoCard.PublishedAt>
      </VideoCard.MediaInfo>
    </VideoCard.Detail>
  </VideoCard>
);

describe('VideoCard', () => {
  it('정상적으로 렌더링됩니다.', () => {
    const { asFragment } = render(VideoCardComponent);

    expect(asFragment()).toMatchSnapshot();
  });
});
