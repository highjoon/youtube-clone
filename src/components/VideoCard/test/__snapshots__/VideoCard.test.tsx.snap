// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard 정상적으로 렌더링됩니다. 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 276px;
  gap: 13px;
  cursor: pointer;
}

.emotion-1 {
  position: relative;
}

.emotion-2 {
  height: 153px;
  background-image: url(https://blog.snappa.com/wp-content/uploads/2020/01/BestYoutubeThumbnailSize_Blogpost.jpg);
  -webkit-background-size: cover;
  background-size: cover;
  -webkit-background-position: center;
  background-position: center;
  background-repeat: no-repeat;
}

.emotion-3 {
  position: absolute;
  right: 5px;
  bottom: 5.5px;
  width: 42px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1px 4px;
  background-color: var(--color-primary-black);
  color: var(--color-primary-white);
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 14px;
}

.emotion-5 {
  display: inline;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  border-radius: 50%;
}

.emotion-7 {
  font-size: 14px;
  color: var(--color-primary-black);
  margin-bottom: 5px;
  white-space: pre-wrap;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
}

.emotion-8 {
  display: block;
  font-size: 14px;
  color: var(--color-black-dark2);
}

.emotion-9 {
  font-size: 14px;
  line-height: 16px;
  color: var(--color-black-dark2);
}

<li
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      />
      <div
        class="emotion-3"
      >
        4:13
      </div>
    </div>
    <div
      class="emotion-4"
    >
      <img
        class="emotion-5"
        src="https://avatars.githubusercontent.com/u/63948484?v=4"
      />
      <div
        class="emotion-6"
      >
        <span
          class="emotion-7"
        >
           The Dark Knight Rises (2012) - Batman's Sacrifice Scene (9/10) Batman's Sacrifice Scene
          (9/10)
        </span>
        <span
          class="emotion-8"
        >
          Movieclips Trailers
        </span>
        <span
          class="emotion-9"
        >
          60M views •
        </span>
        <span
          class="emotion-9"
        >
           1 month ago
        </span>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
